File: src/search/eval.rs
use std::f32::consts::E;

use crate::game::{data::Spin, Game, BOARD_BUFFER, BOARD_HEIGHT};

pub struct Weights {
  pub height: i32,
  pub upper_half_height: i32,
  pub upper_quarter_height: i32,
  pub center_height: i32,

	pub clear_none: i32,
	pub clear_mini: i32,
	pub clear_normal: i32,

  pub sent: i32,

  pub b2b: i32,
  pub combo: i32,

  pub holes: i32,
  pub covered_holes: i32,

  pub unevenness: i32,
}

const HEIGHT: i32 = (BOARD_HEIGHT - BOARD_BUFFER) as i32;

const GENERAL_MULTIPLIER: i32 = 1000;
const GENERAL_MULTIPLIER_F32: f32 = GENERAL_MULTIPLIER as f32;

const WEIGHTS_HANDTUNED: Weights = Weights {
  height: -50,
  upper_half_height: -150,
  upper_quarter_height: -300,
  center_height: -100,

	clear_none: -5,
	clear_mini: 5,
	clear_normal: 10,

  sent: 0,

  b2b: 30,
  combo: 10,

  holes: -13,
  covered_holes: -30,

  unevenness: -3,
};

pub fn eval(state: &Game, sent: u16, clears: Vec<Spin>) -> i32 {
  let mut score = 0;

  // state.print();
  // println!(
  //   "{} {} {} {} {}",
  //   state.board.max_height(),
  //   state.board.center_height(),
  //   sent,
  //   (state.b2b + 1),
  //   state.board.count_holes()
  // );

  score += (state.board.max_height()) * GENERAL_MULTIPLIER / HEIGHT * WEIGHTS_HANDTUNED.height;
  score += (state.board.upper_half_height()) * GENERAL_MULTIPLIER / HEIGHT
    * WEIGHTS_HANDTUNED.upper_half_height;
  score += (state.board.upper_quarter_height()) * GENERAL_MULTIPLIER / HEIGHT
    * WEIGHTS_HANDTUNED.upper_quarter_height;
  score +=
    (state.board.center_height()) * GENERAL_MULTIPLIER / HEIGHT * WEIGHTS_HANDTUNED.center_height;

	for c in clears {
		score += match c {
			Spin::None => WEIGHTS_HANDTUNED.clear_none,
			Spin::Mini => WEIGHTS_HANDTUNED.clear_mini,
			Spin::Normal => WEIGHTS_HANDTUNED.clear_normal,
		} * GENERAL_MULTIPLIER;
	}

  score += sent as i32 * GENERAL_MULTIPLIER * WEIGHTS_HANDTUNED.sent;

  // add 1 so baseline is 0
  score += ((((state.b2b + 1) as f32 + E).ln() - 1.0) * GENERAL_MULTIPLIER_F32) as i32
    * WEIGHTS_HANDTUNED.b2b;
  score += (state.combo as i32 + 1) * GENERAL_MULTIPLIER * WEIGHTS_HANDTUNED.combo;

  score += state.board.count_holes() * GENERAL_MULTIPLIER * WEIGHTS_HANDTUNED.holes;
  score += state.board.covered_holes() * GENERAL_MULTIPLIER * WEIGHTS_HANDTUNED.covered_holes;

  score += state.board.unevenness() * GENERAL_MULTIPLIER * WEIGHTS_HANDTUNED.unevenness;

  score
}

File: ts/search/eval.ts
import { Game, BOARD_BUFFER, BOARD_HEIGHT } from "../game/index.js";
import { Spin } from "../game/data.js";

export interface Weights {
  height: number;
  upperHalfHeight: number;
  upperQuarterHeight: number;
  centerHeight: number;
  extraWells: number;
  clearNone: number;
  clearMini: number;
  clearNormal: number;
  sent: number;
  b2b: number;
  combo: number;
  holes: number;
  coveredHoles: number;
  overstackedHoles: number;
  unevenness: number;
}

export class WeightsData {
  static mutate(weights: Weights, threshold: number, changeFactor: number): Weights {
    const getMultiplier = () => {
      const v = Math.random();
      if (v < threshold) {
        return 0;
      } else if (v < threshold + (1.0 - threshold) / 2.0) {
        return 1;
      } else {
        return -1;
      }
    };

    return {
      height: weights.height + changeFactor * getMultiplier(),
      upperHalfHeight: weights.upperHalfHeight + changeFactor * getMultiplier(),
      upperQuarterHeight: weights.upperQuarterHeight + changeFactor * getMultiplier(),
      centerHeight: weights.centerHeight + changeFactor * getMultiplier(),
      extraWells: weights.extraWells + changeFactor * getMultiplier(),
      clearNone: weights.clearNone + changeFactor * getMultiplier(),
      clearMini: weights.clearMini + changeFactor * getMultiplier(),
      clearNormal: weights.clearNormal + changeFactor * getMultiplier(),
      sent: weights.sent + changeFactor * getMultiplier(),
      b2b: weights.b2b + changeFactor * getMultiplier(),
      combo: weights.combo + changeFactor * getMultiplier(),
      holes: weights.holes + changeFactor * getMultiplier(),
      coveredHoles: weights.coveredHoles + changeFactor * getMultiplier(),
      overstackedHoles: weights.overstackedHoles + changeFactor * getMultiplier(),
      unevenness: weights.unevenness + changeFactor * getMultiplier(),
    };
  }
}

export const WEIGHTS_HANDTUNED: Weights = {
  height: -51,
  upperHalfHeight: -25,
  upperQuarterHeight: -34,
  centerHeight: -100,
  extraWells: -100,
  clearNone: 0,
  clearMini: 100,
  clearNormal: 0,
  sent: 324,
  b2b: 164,
  combo: 191,
  holes: -472,
  coveredHoles: -361,
  overstackedHoles: -81,
  unevenness: -184,
};

export function evaluate(
  game: Game,
  weights: Weights,
  linesSent: number,
  clearTypes: Spin[]
): number {
  let score = 0;

  score += game.board.maxHeight() * weights.height;
  score += game.board.upperHalfHeight() * weights.upperHalfHeight;
  score += game.board.upperQuarterHeight() * weights.upperQuarterHeight;
  score += game.board.centerHeight() * weights.centerHeight;

  score += Math.max(0, game.board.wells() - 1) * weights.extraWells;

  const clearCounts = clearTypes.reduce(
    (counts, spin) => {
      switch (spin) {
        case Spin.None:
          counts.none++;
          break;
        case Spin.Mini:
          counts.mini++;
          break;
        case Spin.Normal:
          counts.normal++;
          break;
      }
      return counts;
    },
    { none: 0, mini: 0, normal: 0 }
  );

  score += clearCounts.none * weights.clearNone;
  score += clearCounts.mini * weights.clearMini;
  score += clearCounts.normal * weights.clearNormal;

  score += linesSent * weights.sent;

  score += game.b2b * weights.b2b;
  score += game.combo * weights.combo;

  score += game.board.countHoles() * weights.holes;
  score += game.board.coveredHoles() * weights.coveredHoles;
  score += game.board.overstackedHoles() * weights.overstackedHoles;

  score += game.board.unevenness() * weights.unevenness;

  if (game.toppedOut()) {
    score -= 1000000;
  }

  if (BOARD_HEIGHT - BOARD_BUFFER < game.board.maxHeight()) {
    score -= Math.pow(game.board.maxHeight() - (BOARD_HEIGHT - BOARD_BUFFER), 3) * 10;
  }

  return score;
}
